src/wallet/wallet.cpp

```diff
@@ -4021,3 +4042,114 @@ bool CMerkleTx::AcceptToMemoryPool(const CAmount& nAbsurdFee, CValidationState&
 {
     return ::AcceptToMemoryPool(mempool, state, tx, true, nullptr, nullptr, false, nAbsurdFee);
 }
+
+bool CWallet::IsSolvable(const CScript& script) const
+{
+    // This check is to make sure that the script we created can actually be solved for and signed by us
+    // if we were to have the private keys. This is just to make sure that the script is valid and that,
+    // if found in a transaction, we would still accept and relay that transaction. In particular,
+    // it will reject witness outputs that require signing with an uncompressed public key.
+    DummySignatureCreator creator(this);
+    SignatureData sigs;
+    return (ProduceSignature(creator, script, sigs) && VerifyScript(sigs.scriptSig, script, &sigs.scriptWitness, MANDATORY_SCRIPT_VERIFY_FLAGS | SCRIPT_VERIFY_WITNESS_PUBKEYTYPE, creator.Checker()));
+}
+
+CKeyID CWallet::GetKeyForDestination(const CTxDestination& dest) const
+{
+    // Only supports P2PKH, P2WPKH, P2SH-P2WPKH.
+    if (boost::get<CKeyID>(&dest)) {
+        return boost::get<CKeyID>(dest);
+    }
+    if (boost::get<WitnessV0KeyHash>(&dest)) {
+        return CKeyID(boost::get<WitnessV0KeyHash>(dest));
+    }
+    if (boost::get<CScriptID>(&dest)) {
+        CScript script;
+        CTxDestination inner_dest;
+        if (GetCScript(boost::get<CScriptID>(dest), script) && ExtractDestination(script, inner_dest)) {
+            if (boost::get<WitnessV0KeyHash>(&inner_dest)) {
+                return CKeyID(boost::get<WitnessV0KeyHash>(inner_dest));
+            }
+        }
+    }
+    return CKeyID();
+}
+
+static const std::string STYLE_STRING_NONE = "";
+static const std::string STYLE_STRING_LEGACY = "legacy";
+static const std::string STYLE_STRING_P2SH = "p2sh";
+static const std::string STYLE_STRING_SEGWIT = "segwit";
+
+OutputStyle ParseStyle(const std::string& style)
+{
+    if (style == STYLE_STRING_LEGACY) {
+        return STYLE_LEGACY;
+    } else if (style == STYLE_STRING_P2SH || style == "default") {
+        return STYLE_P2SH;
```

Comment:
Default should probably remain legacy until support for newer styles is complete (eg, `import*` RPC)?

Reply:
I'd rather not. This makes the existing tests far more powerful.

Reply:
Tests don't need to (and probably shouldn't) rely on defaults...

Reply:
Yes and no.

I think that many tests, which are primarily testing things unrelated to addresses, should use the default address type. They don't strictly rely on them, but it's better if they work with either type, and it's certainly easier to not need to change all of them.